ibc-core/ics04-channel/types/src/timeout/timestamp.rs

```diff
@@ -0,0 +1,145 @@
+/// of a timeout when converting from a zero protobuf value.
+#[cfg_attr(
+    feature = "parity-scale-codec",
+    derive(
+        parity_scale_codec::Encode,
+        parity_scale_codec::Decode,
+        scale_info::TypeInfo
+    )
+)]
+#[cfg_attr(
+    feature = "borsh",
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq)]
+pub enum TimeoutTimestamp {
+    Never,
+    At(Timestamp),
+}
+
+impl TimeoutTimestamp {
+    /// Creates a new timeout timestamp from a given nanosecond value.
+    pub fn from_nanoseconds(nanoseconds: u64) -> Self {
+        Self::from(nanoseconds)
+    }
+
+    /// Returns the timestamp in nanoseconds, where 0 indicates that the absence
+    /// of a timeout.
+    pub fn nanoseconds(&self) -> u64 {
+        match self {
+            Self::At(timestamp) => timestamp.nanoseconds(),
+            Self::Never => 0,
+        }
+    }
+
+    /// Returns `true` if the timeout timestamp is set.
+    pub fn is_set(&self) -> bool {
+        match self {
+            TimeoutTimestamp::At(_) => true,
+            TimeoutTimestamp::Never => false,
+        }
+    }
+
+    /// Returns a timeout timestamp that never expires.
+    pub fn no_timeout() -> Self {
+        Self::Never
+    }
+
+    /// Check if a timestamp is *strictly past* the timeout timestamp, and thus
+    /// is deemed expired.
+    pub fn has_expired(&self, timestamp: &Timestamp) -> bool {
+        match self {
+            Self::At(timeout_timestamp) => timestamp > timeout_timestamp,
+            // When there's no timeout, timestamps are never expired
+            Self::Never => false,
+        }
+    }
+}
+
+impl From<Timestamp> for TimeoutTimestamp {
+    fn from(timestamp: Timestamp) -> Self {
+        TimeoutTimestamp::At(timestamp)
+    }
+}
+
+impl From<u64> for TimeoutTimestamp {
+    fn from(timestamp: u64) -> Self {
+        if timestamp == 0 {
+            TimeoutTimestamp::Never
+        } else {
+            TimeoutTimestamp::At(Timestamp::from_nanoseconds(timestamp))
+        }
+    }
+}
+
+impl Display for TimeoutTimestamp {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self {
+            TimeoutTimestamp::At(timeout_timestamp) => write!(f, "{timeout_timestamp}"),
+            TimeoutTimestamp::Never => write!(f, "no timeout timestamp"),
+        }
+    }
+}
+
+impl Add<Duration> for TimeoutTimestamp {
+    type Output = Result<Self, PacketError>;
+
+    fn add(self, rhs: Duration) -> Self::Output {
+        match self {
+            TimeoutTimestamp::At(timestamp) => {
+                let new_timestamp = timestamp.add(rhs)?;
+                Ok(TimeoutTimestamp::At(new_timestamp))
+            }
+            TimeoutTimestamp::Never => Err(PacketError::MissingTimeout),
```

Comment:
There's a bit of a logical inconsistency here, since a timestamp of 0 gets encoded to the `Never` variant, which can't be added or subtracted. But then that means that we can't add or subtract a timestamp of 0. Just wondering if this might cause issues.

Reply:
There are few subtle details here:

(1) `Timestamp::from_nanoseconds(0)` gives us a date of 1970-01-01T00:00:00Z, on which it's reasonable to be able to perform addition or subtraction. In fact we can have `TimeoutTimestamp::At(Timestamp::from_nanoseconds(0))` which is the same way we used to with the previous `Option<Time>` type as well.

(2) Before, [any addition or subtraction involving `None` always resulted in `None`](https://github.com/cosmos/ibc-rs/blob/5779e868d0330999792d4148452b51f40310182d/ibc-primitives/src/types/timestamp.rs#L239-L240), which didn't seem right to me. Since `None` represents non-existence, there's nothing to add or subtract. so it just should return an error instead.

[9d547d3](https://github.com/cosmos/ibc-rs/pull/1307/commits/9d547d3a6dc2dccbb026359565251c5a26f24f49) I’ve added tests for different possible combinations to make things clearer. Hope those help.

Reply:
Sounds good! The tests are a great addition 🙂